### ISSUE TITLE
Create BigQueryDataset wired to an existing BigQuery dataset

### ISSUE THREAD

**Issue by marcoderama** [GoogleCloudPlatform/k8s-config-connector#327]
Ultimately, I would like a user to be able to write to an existing BigQuery table.  Using GCC's BigQueryTable I can successfully grant a user write access to a BigQuery table, but...

It doesn't work.  When the user tries to POST an update, that fails with
```
  "message" : "Access Denied: Dataset myproject:dev_cluster_data: User does not have bigquery.datasets.get permission for dataset myproject:dev_cluster_data.",
  "status" : "PERMISSION_DENIED"
```
Theoretically this shouldn't be necessary, but see https://issuetracker.google.com/issues/173506894.

In order to get around that, I thought I'd follow the example in issue https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/111#issuecomment-597736636.  That should allow me to grant the required access to the dataset.

I can successfully create a GCC BigQueryDataset object with the access, but I don't see either:
1. an actual BigQuery dataset created in the GCP console, or
1. see any way to associate the GCC BigQueryDataset with an existing actual BigQuery dataset.

And finally I'm down to my question...  How can I associate a GCC BigQueryDataset with an existing GCP BigQuery dataset (so that I can modify access)?  I fear I may be missing something simple and/or not understanding how GCC's BigQueryDataset is supposed to be used.

(And it's not via the `metadata.name` of the GCC object cuz I tried that with no luck.)

**Comment by marcoderama** [comment]
I should add that I'm using GCC 1.31.1.

**Comment by maqiuyujoyce** [comment]
Hi @marcoderama , thank you for your questions and sorry for the frustration you've been through.

> Ultimately, I would like a user to be able to write to an existing BigQuery table. Using GCC's BigQueryTable I can successfully grant a user write access to a BigQuery table, but...

> It doesn't work. When the user tries to POST an update, that fails with
> ```
>   "message" : "Access Denied: Dataset myproject:dev_cluster_data: User does not have bigquery.datasets.get permission for dataset myproject:dev_cluster_data.",
>   "status" : "PERMISSION_DENIED"
> ```

> Theoretically this shouldn't be necessary, but see https://issuetracker.google.com/issues/173506894.

To clarify, were you able to verify that the GCP BigQuery table was successfully created, and the permission had been granted? 
The issue (https://issuetracker.google.com/issues/173506894) seems to be talking about an edge case when the table doesn't exist. If you've run into the same situation, the permission_denied error should be an expected error from the BigQuery service. The user should be able to access normally once the table is created.

If it's a different situation, could you describe the exact scenario when the user encountered this error?
I verified that I can update the table description as the service account after I created the BigQueryTable and granted the `roles/bigquery.dataEditor` role to the service account using Config Connector. Would be great if we understand more about your scenario so that we can reproduce the error.

> In order to get around that, I thought I'd follow the example in issue #111 (comment). That should allow me to grant the required access to the dataset.

> I can successfully create a GCC BigQueryDataset object with the access, but I don't see either:

>  1. an actual BigQuery dataset created in the GCP console, or

Could you share the yaml you've used to create the BigQueryDataset? We'd like to learn how to reproduce it. I created a BigQueryDataset using version 1.13.1 with `access` field, and verified that the GCP BigQuery dataset was created and the permission was granted.


>  2. see any way to associate the GCC BigQueryDataset with an existing actual BigQuery dataset.

> And finally I'm down to my question... How can I associate a GCC BigQueryDataset with an existing GCP BigQuery dataset (so that I can modify access)? I fear I may be missing something simple and/or not understanding how GCC's BigQueryDataset is supposed to be used.

> (And it's not via the metadata.name of the GCC object cuz I tried that with no luck.)

Unfortunately, we have a bug managing list fields (e.g. `access` field in BigQueryDataset) in resources. A similar issue is #313. We are currently working on fixing the bug and will keep you updated. 

Using the `metadata.name` is the right way to do acquisition, and I suspect that it didn't work for you because you specified the `access` field. You can verify it by running `kubectl logs cnrm-controller-manager-0 manager -n cnrm-system` to check the error logs. If you find `"error":"INTERNAL ERROR: panic: observed a panic: expected state's set to have value explicitly provided in spec` in the error logs, then `access` field should be the issue.

For now, a workaround would be [acquiring a BigQuery dataset without the `access` field](https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources#acquiring_a_dataset) and manage the permissions using [other tools (e.g. Google Cloud Console, `bq`, etc)](https://cloud.google.com/bigquery/docs/dataset-access-controls#controlling_access_to_a_dataset).

**Comment by marcoderama** [comment]
To start, I should say that we ended up using other tools/means to do what we needed to (as you suggested at the end of your reply), so I'm not as familiar with the issues using GCC anymore.  I'll try to reproduce them the best I can though.

> To clarify, were you able to verify that the GCP BigQuery table was successfully created, and the permission had been granted?
> The issue (https://issuetracker.google.com/issues/173506894) seems to be talking about an edge case when the table doesn't exist. If you've run into the same situation, the permission_denied error should be an expected error from the BigQuery service. The user should be able to access normally once the table is created.
> 
> If it's a different situation, could you describe the exact scenario when the user encountered this error?
> I verified that I can update the table description as the service account after I created the BigQueryTable and granted the `roles/bigquery.dataEditor` role to the service account using Config Connector. Would be great if we understand more about your scenario so that we can reproduce the error.

I misspoke originally.  Using GCC's IAMPolicyMember I was able to grant the `roles/bigquery.dataEditor` role to a user for a _pre-exising_ BigQuery table.  It was when that user tried to actually write to the table that we got the PERMISSION_DENIED response.  I thought I might get past that by granting the appropriate access to the _pre-existing_ BigQuery dataset.

I've just done some more testing and figured this much out:

Consider this yaml:
```
apiVersion: bigquery.cnrm.cloud.google.com/v1beta1
kind: BigQueryDataset
metadata:
  name: namewithoutanydashes
  annotations:
    cnrm.cloud.google.com/delete-contents-on-destroy: "false"
    cnrm.cloud.google.com/project-id: myproject
spec:
  access:
  - role: WRITER
    # role: bigquery.datasets.get  # <-- try this instead once the more general WRITER works
    userByEmail: marcoderama-test@myproject.iam.gserviceaccount.com
  description: BigQuery dataset
```
* I can successfully create a BigQuery dataset using GCC and that yaml.  Frankly, I have no idea how to tell if the `access` list worked though.  (There must be a `gcloud` way but I couldn't find anything in the GCP Console.)

And now a few buts...

* When I tried to create a GCC BigQueryDataset for a _pre-existing_ dataset, (that is, the `metadata.name` matched the name of an existing dataset) I see the error you mentioned show up in the CNRM operator log.
```
{"severity":"error","logger":"controller-runtime.controller","msg":"Reconciler error","controller":"bigquerydataset-controller","request":"a8afa71b-f9d6-4cf2-8231-d383e40e7ccb/namewithoutanydashes","error":"INTERNAL ERROR: panic: observed a panic: expected state's set to have value explicitly provided in spec
...
```
* If I remove the `access` section of the above yaml, then things work without an error (and I get a Status).
   * The generated `access` section has the three defaults and me as an owner.
   * I don't know how to tell if this is actually associated with the pre-existing BigQuery table of the same name.

* There appears to be a problem with naming and dashes/underscores (-/_)
   * BigQuery datasets (and tables?) cannot include dashes, only underscores
   * GCC's BigQueryDataset allows only dashes and not underscores.

I was originally trying to associate a GCC BigQueryDataset object with an existing dataset that had underscores in its name.  (That's why I asked about the name association originally.)  That undoubtedly added to my confusion.

Summary:
* Some things work but I can't set the dataset access when attaching to an existing dataset.  (Known issue I guess)
* How can I know if a GCC BigQueryDataset object is properly wired to an existing dataset.  (Just a question)
* I don't see how one could attach to an existing, or create a new, dataset with underscores in its name.  The CR verification fails.  When I try, I get
```
The BigQueryDataset "name_with_underscores" is invalid: metadata.name: Invalid value: "name_with_underscores": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character ...
```
* GCC will allow one to specify a name with underscores but that will fail further down the line since BigQuery will reject it.  When I try, GCC happily creates it but the Status shows:
```
    message: 'Update call failed: error applying desired state: summary: Error creating
      Dataset: googleapi: Error 400: Invalid dataset ID "name-with-dashes". Dataset
      IDs must be alphanumeric (plus underscores) and must be at most 1024 characters
      long., invalid, detail: '
```


Thanks for your time on this.

**Comment by maqiuyujoyce** [comment]
Thank you so much for all the details, @marcoderama ! Please find my answers inline.

> * Some things work but I can't set the dataset access when attaching to an existing dataset. (Known issue I guess)

Yep, managing the `access` field is a known issue that we are actively working on (#313). Will keep you updated.

> * How can I know if a GCC BigQueryDataset object is properly wired to an existing dataset? (Just a question)

If the GCP resource already exists with the same resource name as the Config Connector resource, then Config Connector takes control of the resource. So if the Config Connector resource has the same name of the GCP resource, and the status of the Config Connector resource has the message `The resource is up to date`, you can safely assume that the "wiring" has been done successfully. You can find the status by running `kubectl describe bigquerydataset [your-config-connector-bigquerydataset-name]`.

> * I don't see how one could attach to an existing, or create a new, dataset with underscores in its name. The CR verification fails. When I try, I get
> ```
> The BigQueryDataset "name_with_underscores" is invalid: metadata.name: Invalid value: "name_with_underscores": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character ...
> ```
> * GCC will allow one to specify a name with dashes but that will fail further down the line since BigQuery will reject it. When I try, GCC happily creates it but the Status shows:
> ```
>     message: 'Update call failed: error applying desired state: summary: Error creating
>      Dataset: googleapi: Error 400: Invalid dataset ID "name-with-dashes". Dataset
>      IDs must be alphanumeric (plus underscores) and must be at most 1024 characters
>      long., invalid, detail: '
> ```

Yes, that was a limitation Config Connector had before v1.32.0 for BigQuery resources.

At the version you were using, i.e. 1.31.1, you can't use either '-' or '_' in the GCP BigQuery dataset and the Config Connector BigQueryDataset resource in order to wire them up because they have conflicting naming conventions.

But we just released v1.32.0, which supports resource ID overrides using the `resourceID` field. If your existing GCP BigQuery dataset has the name `test_gcp_dataset`, then you can use the follow yaml to acquire it:
```
apiVersion: bigquery.cnrm.cloud.google.com/v1beta1
kind: BigQueryDataset
metadata:
  name: test-dataset
spec:
  resourceID: test_gcp_dataset
```
The detailed documentation will be coming soon.

Regarding
> * I can successfully create a BigQuery dataset using GCC and that yaml. Frankly, I have no idea how to tell if the access list worked though. (There must be a gcloud way but I couldn't find anything in the GCP Console.)

Yes, the GCP Console doesn't use the keyword "access" in the UI, but instead, calls it "Share Dataset". Details can be found [here](https://cloud.google.com/bigquery/docs/dataset-access-controls). And there is actually no `gcloud` way of doing it. I believe BigQuery has it's own CLI so users need to use `bq` instead of `gcloud` to manage GCP BigQuery resources.

**Comment by marcoderama** [comment]
Thanks for informative and comprehensive reply!  You've answered all my questions.

* I'll be upgrading to GCC 1.32.0, but
* I'll have to wait for a resolution to issue https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/313 anyway.
* Thanks for clarifying how to connect to (and know it worked) an existing BQ resource.
* Thanks for info on the new `resourceID` field.
* Thanks for the GCP Console pointer wrt determining access/sharing.  Just what I was looking for.

**Comment by maqiuyujoyce** [comment]
Hi @marcoderama , we've fixed the underlying issue in the 1.33.0 release, and I've verified that I can acquire a BigQueryDataset with `access` field and then update the `access` field successfully. Can you upgrade your version and see if it fixes your issue?

**Comment by maqiuyujoyce** [comment]
Hi @marcoderama I'm closing the issue now considering no response in more than 2 months. Please let us know if you have any followup questions.
